libs/constrained-generators/src/Constrained/Instances.hs

```diff
@@ -67,5 +67,25 @@ instance IsUniverse fn => Functions (BoolFn fn) fn where
       let args = appendList (mapList (\(Value a) -> Lit a) pre) (v' :> mapList (\(Value a) -> Lit a) suf)
        in Let (App (injectFn fn) args) (v :-> ps)
   propagateSpecFun Not (NilCtx HOLE) spec = caseBoolSpec spec (equalSpec . not)
+  propagateSpecFun And (HOLE :? Value (s :: Bool) :> Nil) spec = caseBoolSpec spec (okAnd s)
+  propagateSpecFun And (Value (s :: Bool) :! NilCtx HOLE) spec = caseBoolSpec spec (okAnd s)
+  propagateSpecFun Or (HOLE :? Value (s :: Bool) :> Nil) spec = caseBoolSpec spec (okOr s)
+  propagateSpecFun Or (Value (s :: Bool) :! NilCtx HOLE) spec = caseBoolSpec spec (okOr s)
 
   mapTypeSpec Not _ = typeSpec ()
+
+-- | We have something like ('constant' &&. HOLE) must evaluate to 'need'. Return a (Spec fn Bool) for HOLE, that makes that True.
+okAnd :: Bool -> Bool -> Spec fn Bool
+okAnd constant need = case (constant, need) of
+  (True, True) -> MemberSpec [True]
+  (True, False) -> MemberSpec [False]
+  (False, False) -> TrueSpec
+  (False, True) -> ErrorSpec [" (and_ " ++ show constant ++ " HOLE) must equal True. That cannot be the case."]
+
+-- | We have something like ('constant' ||. HOLE) must evaluate to 'need'. Return a (Spec fn Bool) for HOLE, that makes that True.
+okOr :: Bool -> Bool -> Spec fn Bool
+okOr constant need = case (constant, need) of
```

Comment:
This is not quite correct, c.f. my comments on #4113.

Reply:
OK, I made some changes. Please check that I got it right.